### DISCUSSION TITLE
defaultRegistryUrls has no effect for updates using npm datasource

### DISCUSSION THREAD

**Discussion by rdicroce** [renovatebot/renovate#20989] [Request Help]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

35.7.0

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### If you're self-hosting Renovate, tell us what version of the platform you run.

15.3.3-ee

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

defaultRegistryUrls should not be overridden by the hard-coded default.

### What you tried so far.

I have a self-hosted environment, and I'm trying to configure Renovate such that when it looks for new versions, it uses our internal Nexus server, which proxies all of the relevant datasources. This is mostly working, except for the yarn engine defined in package.json.

I've created a minimal reproduction [here](https://github.com/rdicroce/renovate-npm-url-issue) that demonstrates the problem. After looking at the code, I think I've also figured out the cause, which I explained in the readme.

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: 1 flattened updates found: yarn
DEBUG: Returning 1 branch(es)
DEBUG: config.repoIsOnboarded=true
DEBUG: packageFiles with updates
{
  "baseBranch": "master",
  "config": {
    "npm": [
      {
        "deps": [
          {
            "depType": "engines",
            "depName": "yarn",
            "currentValue": "3.1.1",
            "datasource": "npm",
            "commitMessageTopic": "Yarn",
            "packageName": "@yarnpkg/cli",
            "prettyDepType": "engine",
            "updates": [
              {
                "bucket": "non-major",
                "newVersion": "3.4.1",
                "newValue": "3.4.1",
                "releaseTimestamp": "2023-02-01T16:14:39.479Z",
                "newMajor": 3,
                "newMinor": 4,
                "updateType": "minor",
                "branchName": "renovate/yarn-3.x"
              }
            ],
            "warnings": [],
            "versioning": "npm",
            "sourceUrl": "https://github.com/yarnpkg/berry",
            "registryUrl": "https://registry.npmjs.org",
            "sourceDirectory": "packages/yarnpkg-cli",
            "currentVersion": "3.1.1",
            "isSingleVersion": true,
            "fixedVersion": "3.1.1"
          }
        ],
        "packageFileVersion": "0.0.1",
        "managerData": {
          "packageJsonName": "foobar",
          "yarnZeroInstall": false,
          "hasPackageManager": false
        },
        "skipInstalls": true,
        "packageFile": "package.json",
        "constraints": {
          "yarn": "3.1.1"
        }
      }
    ]
  }
}
```

</details>

**Comment by rarkins** [comment]
npm is unfortunately a bit unique in the way it does things. Have you tried configuration like `"nmprc": "registry=https://registry.yarnpkg.com"` in your `renovate.json`?

**Reply by rdicroce** [reply]
That does get it working. If you're not planning to fix this, I'd suggest adding a warning about it to the docs for the npm datasource.

**Reply by rarkins** [reply]
PRs welcome
